solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -2563,6 +2581,11 @@ public PlacementPluginFactory<? extends PlacementPluginConfig> getPlacementPlugi
     return this.distributedCollectionCommandRunner;
   }
 
+  /** Returns default http solr client. */
+  public SolrClient getDefaultHttpClient() {
```

Comment:
Can return the specific type; there are async methods there which can be useful.

Reply:
We could recreate Http2SolrClient without casting, but I aimed to follow the principle of using base classes to allow for future implementation changes without altering the method signature. If requestAsync isn't used often, we could cast it when needed. or, Would that make it less clean?

Reply:
Since this is specifically an HttpClient (it's in the method name!) I think it's totally reasonable to return this type.  Casting is sad; it's a code smell.  reqeustAsync is somewhat new so you want see it called too many times.  It remains to be seen what the solution to [SOLR-17256](https://issues.apache.org/jira/browse/SOLR-17256) is but I could imagine a new request method containing the URL; if so then yet another reason to return a specific type.

Reply:
I was thinking about this one, I believe sticking to the parent class would benefit us in a long, I mean returning SolrClient. Instead we should introduce the empty/abstract method `requestAsync` in SolrClient, and override in all the sub classes, and where it does not support just throw an exception.

Reply:
Okay for requestAsync even if debatable but what about SOLR-17256?  The solution for SOLR-17256 is not determined yet but the likely solution IMO would result in a URL specific request method on the client.  That doesn't belong on SolrClient.

Reply:
What If we introduced an requestAsyn with an empty body instead of making it abstract in the SolrClient that way It would not enforce others to implement it.

Reply:
I've observed that some classes take implementation classes as references, which makes them harder to change. For example, SolrClientCache initially used the Apache Http Client, but when Http2SolrClient was introduced, a new constructor was needed. Similarly, SolrClientCloudManager uses CloudLegacySolrClient, which complicates future changes. We might eventually need to introduce a SolrClientCloudManager2 to accommodate CloudSolrClient.

Reply:
_If need be_ we can create an interface like `HttpSolrClient` (we have 3 impls) but the only backend impl we care about is Http2SolrClient (jetty) and I'm skeptical we're going to see another migration.

BTW sorry but a no-op/empty requestAsync would be nasty... think of it:  You call a method and it does absolutely nothing when it should have issues the request!

Reply:
I got it! That is ugly. Anyways I am okay with this method returning Http2SolrClient. 

@gerlowskija I already worked on changing it for two files PKIAuth and ReindexCollectionCmd, however we decided not to in this one and deal in separate PR. Take a look at this one, If everything's looking good then I will merge this one first and then I will make another.

Reply:
Awesome, thanks for the heads up @iamsanjay !